File: iot/amqp/terminus/AMQPSource.py
import iot.amqp.avps.AMQPType as AMQPType
import iot.amqp.avps.DistributionMode as DistributionMode
import iot.amqp.avps.TerminusDurability as TerminusDurability
import iot.amqp.avps.TerminusExpiryPolicy as TerminusExpiryPolicy
import iot.amqp.constructor.DescribedConstructor as DescribedConstructor
import iot.amqp.header.api.AMQPUnwrapper as AMQPUnwrapper
import iot.amqp.header.api.AMQPWrapper as AMQPWrapper
import iot.amqp.header.api.HeaderFactoryOutcome as HeaderFactoryOutcome
import iot.amqp.wrappers.AMQPSymbol as AMQPSymbol
import iot.amqp.tlv.impl.TLVList as TLVList
import iot.amqp.tlv.impl.TLVFixed as TLVFixed

class amqpSource():
    def __init__(self,address,durable,expiryPeriod,timeout,dynamic,dynamicNodeProperties,distributionMode,filter,defaultOutcome,outcomes,capabilities):
        self.address = address
        self.durable = durable
        self.expiryPeriod = expiryPeriod
        self.timeout = timeout
        self.dynamic = dynamic
        self.dynamicNodeProperties = dynamicNodeProperties
        self.distributionMode = distributionMode
        self.filter = filter
        self.defaultOutcome = defaultOutcome
        self.outcomes = outcomes
        self.capabilities = capabilities
        self.index = 0
        self.amqpType = AMQPType.amqpType()

    def toArgumentsList(self):
        list = TLVList.tlvList(None,None)
        wrapper = AMQPWrapper.amqpWrapper()
        if self.address is not None:
            list.addElement(0, wrapper.wrap(self.address))
        if self.durable is not None:
            list.addElement(1, wrapper.wrap(self.durable))
        if self.expiryPeriod is not None:
            list.addElement(2, wrapper.wrap(AMQPSymbol.amqpSymbol(self.expiryPeriod.value)))
        if self.timeout is not None:
            list.addElement(3, wrapper.wrap(self.timeout))
        if self.dynamic is not None:
            list.addElement(4, wrapper.wrap(self.dynamic))
        if self.dynamicNodeProperties is not None and isinstance(self.dynamicNodeProperties,dict):
            if self.dynamic is not None:
                if self.dynamic:
                    list.addElement(5, wrapper.wrapMap(self.dynamicNodeProperties))
                else:
                    print("Source's dynamic-node-properties can't be specified when dynamic flag is false")
            else:
                print("Source's dynamic-node-properties can't be specified when dynamic flag is not set")

        if self.distributionMode is not None:
            list.addElement(6, wrapper.wrap(AMQPSymbol.amqpSymbol(self.distributionMode.value)))
        if self.filter is not None and isinstance(self.filter, dict):
            list.addElement(7, wrapper.wrapMap(self.filter))
        if self.defaultOutcome is not None:
            list.addElement(8, self.defaultOutcome.toArgumentsList())
        if self.outcomes is not None and len(self.outcomes) > 0:
            list.addElement(9, wrapper.wrapArray(self.outcomes))
        if self.capabilities is not None and len(self.capabilities) > 0:
            list.addElement(10, wrapper.wrapArray(self.capabilities))

        constructor = DescribedConstructor.describedConstructor(list.getCode(),TLVFixed.tlvFixed(self.amqpType.getValueByKey('SMALL_ULONG'), 0x28))
        list.setConstructor(constructor)
        return list

    def fromArgumentsList(self, list):
        unwrapper = AMQPUnwrapper.amqpUnwrapper()
        if isinstance(list, TLVList):
            if len(list.getList()) > 0 :
                element = list.getList()[0]
                if element is not None and not element.isNull():
                    self.address = unwrapper.unwrapString(element)
            if len(list.getList()) > 1 :
                element = list.getList()[1]
                if element is not None and not element.isNull():
                    self.durable = TerminusDurability.terminusDurability(unwrapper.unwrapUInt(element))
            if len(list.getList()) > 2 :
                element = list.getList()[2]
                if element is not None and not element.isNull():
                    self.expiryPeriod = TerminusExpiryPolicy.terminusExpiryPolicy(unwrapper.unwrapSymbol(element))
            if len(list.getList()) > 3 :
                element = list.getList()[3]
                if element is not None and not element.isNull():
                    self.timeout = unwrapper.unwrapUInt(element)
            if len(list.getList()) > 4 :
                element = list.getList()[4]
                if element is not None and not element.isNull():
                    self.dynamic = unwrapper.unwrapBool(element)
            if len(list.getList()) > 5 :
                element = list.getList()[5]
                if element is not None and not element.isNull():
                    if self.dynamic is not None:
                        if self.dynamic:
                            self.dynamicNodeProperties = unwrapper.unwrapMap(element)
                        else:
                            print("Received malformed Source: dynamic-node-properties can't be specified when dynamic flag is false")
                    else:
                        print("Received malformed Source: dynamic-node-properties can't be specified when dynamic flag is not set")

            if len(list.getList()) > 6 :
                element = list.getList()[6]
                if element is not None and not element.isNull():
                    self.distributionMode = DistributionMode.distributionMode(unwrapper.unwrapSymbol(element))
            if len(list.getList()) > 7:
                element = list.getList()[7]
                if element is not None and not element.isNull():
                    self.filter = unwrapper.unwrapMap(element)

            if len(list.getList()) > 8:
                element = list.getList()[8]
                if element is not None and not element.isNull():
                    code = element.getCode()
                    if code not in (self.amqpType.getValueByKey('LIST_0'),self.amqpType.getValueByKey('LIST_8'),self.amqpType.getValueByKey('LIST_32')):
                        print('Expected type OUTCOME - received: ' + element.getCode())


                self.defaultOutcome = HeaderFactoryOutcome.headerFactoryOutcome.getOutcome(element)
                self.defaultOutcome.fromArgumentsList(element)

            if len(list.getList()) > 9:
                element = list.getList()[9]
                if element is not None and not element.isNull():
                    self.outcomes = unwrapper.unwrapArray(element)
            if len(list.getList()) > 10:
                element = list.getList()[10]
                if element is not None and not element.isNull():
                    self.capabilities = unwrapper.unwrapArray(element)

    def toString(self):
        return 'AMQPSource [address=' + str(self.address) + ', durable=' + str(self.durable) + ', expiryPeriod=' + str(self.expiryPeriod) + ', timeout=' + str(self.timeout) + ', dynamic=' + str(self.dynamic) + ', dynamicNodeProperties=' + str(self.dynamicNodeProperties) + ', distributionMode=' + str(self.distributionMode) + ', filter=' + str(self.filter) + ', defaultOutcome=' + str(self.defaultOutcome) + ', outcomes=' + str(self.outcomes) + ', capabilities=' + str(self.capabilities) + ']'

    def getAddress(self):
        return self.address

    def setAddress(self, address):
        self.address = address

    def getDurable(self):
        return self.durable

    def setDurable(self, durable):
        self.durable = durable

    def getExpiryPeriod(self):
        return self.expiryPeriod

    def setExpiryPeriod(self, expiryPeriod):
        self.expiryPeriod = expiryPeriod

    def getTimeout(self):
        return self.timeout

    def setTimeout(self, timeout):
        self.timeout = timeout

    def getDynamic(self):
        return self.dynamic

    def setDynamic(self, dynamic):
        self.dynamic = dynamic

    def getDynamicNodeProperties(self):
        return self.dynamicNodeProperties

    def setDynamicNodeProperties(self, dynamicNodeProperties):
        self.dynamicNodeProperties = dynamicNodeProperties

    def getDistributionMode(self):
        return self.distributionMode

    def setDistributionMode(self, distributionMode):
        self.distributionMode = distributionMode

    def getFilter(self):
        return self.filter

    def setFilter(self, filter):
        self.filter = filter

    def getDefaultOutcome(self):
        return self.defaultOutcome

    def setDefaultOutcome(self, defaultOutcome):
        self.defaultOutcome = defaultOutcome

    def getOutcomes(self):
        return self.outcomes

    def setOutcomes(self, outcomes):
        self.outcomes = outcomes

    def getCapabilities(self):
        return self.capabilities

    def setCapabilities(self, capabilities):
        self.capabilities = capabilities

File: IoTbrokerClient/src/iot-protocols/amqp/classes/terminus/amqpsource.cpp
/**
 * Mobius Software LTD
 * Copyright 2015-2018, Mobius Software LTD
 *
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */

#include "amqpsource.h"
#include "iot-protocols/amqp/classes/wrappers/amqpsymbol.h"
#include "iot-protocols/amqp/classes/constructors/amqpdescribedconstructor.h"
#include "iot-protocols/amqp/classes/tlv/api/amqptlvfactory.h"
#include "iot-protocols/amqp/classes/tlv/fixed/amqptlvfixed.h"
#include "iot-protocols/amqp/classes/headerapi/amqpwrapper.h"
#include "iot-protocols/amqp/classes/headerapi/amqpunwrapper.h"
#include "iot-protocols/amqp/classes/headerapi/amqpfactory.h"

AMQPSource::AMQPSource()
{
    this->expiryPeriod = NULL;
    this->distributionMode = NULL;
    this->defaultOutcome = NULL;
    this->dynamicNodeProperties = NULL;
    this->filter = NULL;
    this->outcomes = NULL;
    this->capabilities = NULL;
}

AMQPTLVList *AMQPSource::getList()
{
    AMQPTLVList *list = new AMQPTLVList();

    if (!this->address.isEmpty()) {
        list->addElementWithIndex(0, AMQPWrapper::wrapString(this->address));
    }
    if (this->durable != NULL) {
        list->addElementWithIndex(1, AMQPWrapper::wrapUInt(this->durable->getValue()));
    }
    if (this->expiryPeriod != NULL) {
        AMQPSymbol *symbol = new AMQPSymbol(this->expiryPeriod->getName());
        list->addElementWithIndex(2, AMQPWrapper::wrapSymbol(symbol));
    }
    if (this->timeout != NULL) {
        list->addElementWithIndex(3, AMQPWrapper::wrapUInt(this->timeout));
    }
    if (this->dynamic != NULL) {
        list->addElementWithIndex(4, AMQPWrapper::wrapBool(this->dynamic));
    }
    if (this->dynamicNodeProperties.count() != 0) {
        if (this->dynamic != NULL) {
            if ((bool) this->dynamic == true) {
                list->addElementWithIndex(5, AMQPWrapper::wrapMap(this->dynamicNodeProperties));
            } else {
                printf("AMQPSource::getList::dynamicNodeProperties");
            }
        } else {
            printf("AMQPSource::getList::dynamic");
        }
    }
    if (this->distributionMode != NULL) {
        AMQPSymbol *symbol = new AMQPSymbol(this->distributionMode->getName());
        list->addElementWithIndex(6, AMQPWrapper::wrapSymbol(symbol));
    }
    if (this->filter.count() != 0) {
        list->addElementWithIndex(7, AMQPWrapper::wrapMap(this->filter));
    }
    if (this->defaultOutcome != NULL) {
        list->addElementWithIndex(8, this->defaultOutcome->getList());
    }
    if (this->outcomes.count() != 0) {
        list->addElementWithIndex(9, AMQPWrapper::wrapArray(this->outcomes));
    }
    if (this->capabilities.count() != 0) {
        list->addElementWithIndex(9, AMQPWrapper::wrapList(this->capabilities));
    }

    unsigned char * data = NULL;
    data = malloc(sizeof(char));
    data[0] = (char) 0x28;
    AMQPType *type = new AMQPType(AMQP_SMALL_ULONG_TYPE);
    AMQPTLVFixed *fixed = new AMQPTLVFixed(type, data);

    AMQPType *constructorType = new AMQPType(list->getTypeValue());
    AMQPDescribedConstructor *constructor = new AMQPDescribedConstructor(constructorType, fixed);

    list->setConstructor(constructor);

    return list;
}

void AMQPSource::fill(AMQPTLVList *list)
{
    if (list->getList().size() > 0) {
        TLVAMQP *element = list->getList()[0];
        if (!element->isNull()) {
            this->address = AMQPUnwrapper::unwrapString(element);
        }
    }
    if (list->getList().size() > 1) {
        TLVAMQP *element = list->getList()[1];
        if (!element->isNull()) {
            int number = AMQPUnwrapper::unwrapUInt(element);
            this->durable = new AMQPTerminusDurability(number);
        }
    }
    if (list->getList().size() > 2) {
        TLVAMQP *element = list->getList()[2];
        if (!element->isNull()) {
            AMQPSymbol *symbol = AMQPUnwrapper::unwrapSymbol(element);
            AMQPTerminusExpiryPolicy *policy = new AMQPTerminusExpiryPolicy();
            policy = new AMQPTerminusExpiryPolicy(policy->EnumObject::getValue(symbol->getValue()));
            this->expiryPeriod = policy;
        }
    }
    if (list->getList().size() > 3) {
        TLVAMQP *element = list->getList()[3];
        if (!element->isNull()) {
            int number = AMQPUnwrapper::unwrapUInt(element);
            JsonVariant variant;
            variant.set(number);
            this->timeout = variant;
        }
    }
    if (list->getList().size() > 4) {
        TLVAMQP *element = list->getList()[4];
        if (!element->isNull()) {
            int number = AMQPUnwrapper::unwrapBool(element);
            JsonVariant variant;
            variant.set(number);
            this->dynamic = new variant(number);
        }
    }
    if (list->getList().size() > 5) {
        TLVAMQP *element = list->getList()[5];
        if (!element->isNull()) {
            if (this->dynamic != NULL) {
                if ((bool) this->dynamic) {
                    this->dynamicNodeProperties = AMQPUnwrapper::unwrapMap(element);
                } else {
                    printf("AMQPSource::fill::dynamicNodeProperties");
                }
            } else {
                printf("AMQPSource::fill::dynamic");
            }
        }
    }
    if (list->getList().size() > 6) {
        TLVAMQP *element = list->getList()[6];
        if (!element->isNull()) {
            AMQPSymbol *symbol = AMQPUnwrapper::unwrapSymbol(element);
            AMQPDistributionMode *mode = new AMQPDistributionMode();
            mode = new AMQPDistributionMode(mode->EnumObject::getValue(symbol->getValue()));
        }
    }
    if (list->getList().size() > 7) {
        TLVAMQP *element = list->getList()[7];
        if (!element->isNull()) {
            this->filter = AMQPUnwrapper::unwrapMap(element);
        }
    }
    if (list->getList().size() > 8) {
        TLVAMQP *element = list->getList()[8];
        if (!element->isNull()) {
            AMQPType *code = new AMQPType(element->getTypeValue());
            if (code->getValue() != AMQP_LIST0_TYPE && code->getValue() != AMQP_LIST8_TYPE && code->getValue() != AMQP_LIST32_TYPE) {
                printf("AMQPSource::fill::wrong_type");
            }
            this->defaultOutcome = AMQPFactory::outcome((AMQPTLVList *)element);
            this->defaultOutcome->fill((AMQPTLVList *)element);
        }
    }
    if (list->getList().size() > 9) {
        TLVAMQP *element = list->getList()[9];
        if (!element->isNull()) {
            this->outcomes = AMQPUnwrapper::unwrapArray(element);
        }
    }
    if (list->getList().size() > 10) {
        TLVAMQP *element = list->getList()[10];
        if (!element->isNull()) {
            this->capabilities = AMQPUnwrapper::unwrapList(element);
        }
    }
}

void AMQPSource::addDynamicNodeProperties(String key, JsonVariant *value)
{
    JsonVariant *keyVariant;
    keyVariant.set(AMQPSimpleType::symbolToVariant(new AMQPSymbol(key)));
    this->dynamicNodeProperties.insert(keyVariant, value);
}

void AMQPSource::addFilter(String key, JsonVariant *value)
{
    JsonVariant *keyVariant;
    keyVariant.set(AMQPSimpleType::symbolToVariant(new AMQPSymbol(key)));
    this->filter.insert(keyVariant, value);
}

void AMQPSource::addOutcomes(list<String> array)
{
    std::list<String *>::iterator it;
    for(it=array.begin(); it!=array.end(); ++it){
        JsonVariant *itemVariant;
        itemVariant.set(AMQPSimpleType::symbolToVariant(new AMQPSymbol(it)));
        this->outcomes.append(itemVariant);
    }
}

void AMQPSource::addCapabilities(list<String> array)
{
    std::list<String *>::iterator it;
    for(it=array.begin(); it!=array.end(); ++it){
        JsonVariant *itemVariant;
        itemVariant.set(AMQPSimpleType::symbolToVariant(new AMQPSymbol(it)));
        this->capabilities.append(itemVariant);
    }
}

String AMQPSource::getAddress() const
{
    return address;
}

void AMQPSource::setAddress(const String &value)
{
    address = value;
}

AMQPTerminusDurability *AMQPSource::getDurable() const
{
    return durable;
}

void AMQPSource::setDurable(AMQPTerminusDurability *value)
{
    durable = value;
}

AMQPTerminusExpiryPolicy *AMQPSource::getExpiryPeriod() const
{
    return expiryPeriod;
}

void AMQPSource::setExpiryPeriod(AMQPTerminusExpiryPolicy *value)
{
    expiryPeriod = value;
}

JsonVariant *AMQPSource::getTimeout() const
{
    return timeout;
}

void AMQPSource::setTimeout(JsonVariant *value)
{
    timeout = value;
}

JsonVariant *AMQPSource::getDynamic() const
{
    return dynamic;
}

void AMQPSource::setDynamic(JsonVariant *value)
{
    dynamic = value;
}

std::map<JsonVariant *, JsonVariant *> AMQPSource::getDynamicNodeProperties() const
{
    return dynamicNodeProperties;
}

void AMQPSource::setDynamicNodeProperties(const std::map<JsonVariant *, JsonVariant *> &value)
{
    dynamicNodeProperties = value;
}

AMQPDistributionMode *AMQPSource::getDistributionMode() const
{
    return distributionMode;
}

void AMQPSource::setDistributionMode(AMQPDistributionMode *value)
{
    distributionMode = value;
}

std::map<JsonVariant *, JsonVariant *> AMQPSource::getFilter() const
{
    return filter;
}

void AMQPSource::setFilter(const std::map<JsonVariant *, JsonVariant *> &value)
{
    filter = value;
}

AMQPOutcome *AMQPSource::getDefaultOutcome() const
{
    return defaultOutcome;
}

void AMQPSource::setDefaultOutcome(AMQPOutcome *value)
{
    defaultOutcome = value;
}

std::list<JsonVariant *> AMQPSource::getOutcomes() const
{
    return outcomes;
}

void AMQPSource::setOutcomes(const std::list<JsonVariant *> &value)
{
    outcomes = value;
}

std::list<JsonVariant *> AMQPSource::getCapabilities() const
{
    return capabilities;
}

void AMQPSource::setCapabilities(const std::list<JsonVariant *> &value)
{
    capabilities = value;
}
